evm/src/executor/inspector/cheatcodes/ext.rs

```diff
@@ -160,13 +160,43 @@ fn get_env(key: &str, r#type: ParamType, delim: Option<&str>) -> Result<Bytes, B
         .map_err(|e| e.into())
 }
 
-fn read_file(path: &str) -> Result<Bytes, Bytes> {
+macro_rules! check_path_allowed {
+    ($state:ident, $path:ident) => {
+        // Get parent directory of the file that user wants to read/write.
+        let directory = Path::new($path)
+            .parent()
+            .ok_or("Can't get parent directory for the provided path".to_string().encode())?;
+
+        // Check it's a directory and it exists.
+        if !directory.is_dir() {
+            return Err("Parent directory of the file is not a directory. Is it exist?"
+                .to_string()
+                .encode()
+                .into())
+        }
+
+        // Get an absolute path for the directory.
+        // We can't do it with the file path because `fs::canonicalize` requires the file to exist.
+        let path = fs::canonicalize(directory).map_err(|err| err.to_string().encode())?;
+
+        // Check if any of the allowed paths is the prefix for our directory path.
+        if !$state.config.allowed_paths.iter().any(|allowed_path| path.starts_with(allowed_path)) {
+            return Err("Path is not allowed.".to_string().encode().into())
+        }
+    };
```

Comment:
I wonder if we should rather do `root.join(dir)` because this will pick the cwd which may not the project root

Reply:
yeah, that's fair. I think it should also be applied to how we read/write files because afaik currently it's relative to directory where the `forge test` was called which may not be the project root.